frontend/src/main/scala/bloop/exec/JvmProcessForker.scala

```diff
@@ -0,0 +1,163 @@
+      debugSession: Option[DebugSession]
+  ): JvmProcessForker = {
+    val underlying = JvmProcessForker(javaEnv, classpath)
+    debugSession
+      .map(session => new JvmDebuggingForker(underlying, session))
+      .getOrElse(underlying)
+  }
+}
+
+/**
+ * @param env   The configuration describing how to start the new JVM
+ * @param classpath Classpath with which the code should be executed
+ */
+final class JvmForker(env: JavaEnv, classpath: Array[AbsolutePath]) extends JvmProcessForker {
+
+  /**
+   * Creates a `ClassLoader` from the classpath of this `ForkProcess`
+   *
+   * @param parent A parent classloader
+   * @return A classloader constructed from the classpath of this `ForkProcess`
+   */
+  override def newClassLoader(parent: Option[ClassLoader]): ClassLoader = {
+    val classpathEntries = classpath.map(_.underlying.toUri.toURL)
+    new URLClassLoader(classpathEntries, parent.orNull)
+  }
+
+  override def runMain(
+      cwd: AbsolutePath,
+      mainClass: String,
+      args: Array[String],
+      jargs: Array[String],
+      logger: Logger,
+      opts: CommonOptions,
+      extraClasspath: Array[AbsolutePath]
+  ): Task[Int] = {
+    val jvmOptions = jargs ++ env.javaOptions
+    val fullClasspath = (classpath ++ extraClasspath).map(_.syntax).mkString(pathSeparator)
+    val java = env.javaHome.resolve("bin").resolve("java")
+    val classpathOption = "-cp" :: fullClasspath :: Nil
+    val appOptions = mainClass :: args.toList
+    val cmd = java.syntax :: jvmOptions.toList ::: classpathOption ::: appOptions
+
+    val logTask =
+      if (logger.isVerbose) {
+        val debugOptions =
+          s"""
+             |Fork options:
+             |   command      = '${cmd.mkString(" ")}'
+             |   cwd          = '$cwd'""".stripMargin
+        Task(logger.debug(debugOptions)(DebugFilter.All))
+      } else Task.unit
+    logTask.flatMap(_ => Forker.run(cwd, cmd, logger, opts))
+  }
+}
+
+final class JvmDebuggingForker(underlying: JvmProcessForker, debugSession: DebugSession)
+    extends JvmProcessForker {
+
+  override def newClassLoader(parent: Option[ClassLoader]): ClassLoader =
+    underlying.newClassLoader(parent)
+
+  override def runMain(
+      cwd: AbsolutePath,
+      mainClass: String,
+      args: Array[String],
+      jArgs: Array[String],
+      ignored: Logger, // we replace it with the DAP logger TODO this logger does not belong to this param list
+      opts: CommonOptions,
+      extraClasspath: Array[AbsolutePath]
+  ): Task[Int] = {
+    val address = findFreeSocketAddress
+    val jvmOptions = jArgs :+ jdiOpt(address.getHostName, address.getPort)
+    val logger = new DebugSessionLogger(debugSession)
+    val task = underlying.runMain(cwd, mainClass, args, jvmOptions, logger, opts, extraClasspath)
+
+    task
+  }
+
+  private def findFreeSocketAddress: InetSocketAddress = {
+    val server = new ServerSocket(0)
+    server.close()
+    server.getLocalSocketAddress
```

Comment:
So we open a socket and close it to find a free address? I don't think it's the best solution. Whathever runs should find and reserve that socket itself. You can have a race here, no?